### DISCUSSION TITLE
General question regarding dotnet, msbuild and csc

### DISCUSSION THREAD

**Discussion by awa5114** [dotnet/runtime#70715] [Q&A] [answered]
My current understanding of how the .NET 6.0 CLI builds a C# assembly is as follows:

- `dotnet run` calls `msbuild`
- `msbuild` calls `csc`
- `csc` compiles the sources files, using the necessary references along the way

So my impression is that, under the hood, the compiler is *still* `csc.exe`, even when I'm using the higher level `dotnet` CLI. Is my impression correct? Is it worth spending a little time getting more familiar with the intricate details of `csc` for a novice C# programmer or is `csc.exe` going to be deprecated in the (near) future?

**Comment by huoyaoyuan** [comment, marked as answer]
> So my impression is that, under the hood, the compiler is _still_ `csc.exe`, even when I'm using the higher level `dotnet` CLI. Is my impression correct?

Strictly speaking ... it's almost correct.

There are typically two copies of compilers installed. One is at `C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Current\Bin\Roslyn`. As you can see, this copy is used by Visual Studio, and contains .NET Framework entry points.
The other one is at `C:\Program Files\dotnet\sdk\6.0.300\Roslyn\bincore`. This one is used from command line. It contains .NET Core entry points. When executing, it's actually doing `dotnet.exe csc.dll`.

> Is it worth spending a little time getting more familiar with the intricate details of `csc` for a novice C# programmer or is `csc.exe` going to be deprecated in the (near) future?

Neither. `csc` is just one of the entry points of the compiler. It parses arguments from the command line, and does one-time compilation work. Another entry point is `VBCSCompiler`, which keeps the process alive for a long time, and can accept multiple tasks through RPC. `dotnet` CLI also often uses `VBCSCompiler` to reduce the cost of creating process repeatedly.
`csc` itself doesn't contain any detail at all. It just trivially parses the command line. It's actual code body is at https://github.com/dotnet/roslyn/tree/main/src/Compilers/Shared.

The code doing the actual compilation work is in `Microsoft.CodeAnalysis.CSharp.dll` and `Microsoft.CodeAnalysis.dll`. Yes, the name of the compiler is "code analysis". Here is their source code: https://github.com/dotnet/roslyn/tree/main/src/Compilers/CSharp/Portable and https://github.com/dotnet/roslyn/tree/main/src/Compilers/Core/Portable.

**Reply by awa5114** [reply]
Thanks for your informative answer. It turns out I actually have another `csc.exe` file in my machine:
```
PS > where.exe csc
C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Current\Bin\Roslyn\csc.exe
C:\Windows\Microsoft.NET\Framework64\v4.0.30319\csc.exe
```

I have no idea where the second one comes from, but I suspect it has been left there by some previous installation. Any ideas on how I can get rid of that?

**Reply by huoyaoyuan** [reply]
> I have no idea where the second one comes from, but I suspect it has been left there by some previous installation. Any ideas on how I can get rid of that?

As its directory hints, it's part of .NET Framework SDK. If you pass `/?` to it you can see the message:
```
Microsoft (R) Visual C# Compiler version 4.8.4084.0
for C# 5
Copyright (C) Microsoft Corporation. All rights reserved.

This compiler is provided as part of the Microsoft (R) .NET Framework,
but only supports language versions up to C# 5, which is no longer the latest version.
For compilers that support newer versions of the C# programming language, see http://go.microsoft.com/fwlink/?LinkID=533240
```

It was written in C++, unlike Roslyn which is in C#. It's just a legacy component and will be there as-is, together with .NET Framework.

**Reply by PathogenDavid** [reply]
> I have no idea where the second one comes from, but I suspect it has been left there by some previous installation. Any ideas on how I can get rid of that?

That's the legacy compiler that was distributed as a part of the old .NET Framework which is distributed as a part of Windows. You shouldn't use it because it's utterly ancient, but you also shouldn't try to remove it because it's a system component.

> Is it worth spending a little time getting more familiar with the intricate details of csc for a novice C# programmer

I think it's worthwhile just to get familiar with what feeds into the compiler for the sake of curiosity, but you probably should not expect to build projects by invoking it manually as there's quite a lot of heavy lifting being done by the SDK.

If you want to see the `csc` command line that is being invoked by MSBuild, you can turn up the verbosity: `dotnet build --no-incremental --verbosity normal`.

You can also use the [MSBuild Structured Log Viewer](https://github.com/KirillOsenkov/MSBuildStructuredLog) which shows that and a whole lot more (search for `$target CoreCompile`.)
